[Shipping Labels] Migrate config cache from DataStore to Room

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-726

> [!NOTE]
> Don't be alarmed by the size of changes; the biggest change is just the DB schema file, the rest of changes should be manageable.

### Description
As discussed before, this PR migrates the cache storage of `config` data from DataStore to Room. It adds two tables: 
- `WooShippingLabelEntity` for caching shipping labels.
- `WooShippingShipmentEntity` for caching shipments.

For easier review, the changes are grouped into consecutive commits and can be divided as follows:

- **Shipments**: From commit a0a202d to commit 07a9f02
- **Labels**: From commit b00d809 to commit f1fa2ad
- **Unit tests**: commit 8b837b2

### Steps to reproduce
#### Shipments
1. Open shipping label creation form for an order with multiple items.
2. Test splitting shipments.
3. Test leaving the order and coming back, and confirming the shipments are loaded correctly.
4. Test splitting shipments on the web and confirming the changes are loaded correctly.

#### Labels
1. Test shipping label purchase and confirm the correct data is shown after purchase.
2. Test shipping label refund.
3. Test displaying previously purchased labels.

### Testing information
- Confirm that there is no regression in all the above flows.

### The tests that have been performed
The above.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Rename class

The old name was very generic

hichamboushaba committed:
Introduce a new DB table for caching shipments

hichamboushaba committed:
Introduce Dao for shipments

hichamboushaba committed:
Bump the DB version and include shipments

hichamboushaba committed:
Update shipping labels form to use Room for caching shipments

hichamboushaba committed:
Add a new table for caching shipping labels

hichamboushaba committed:
Stop caching shipments data to DataStore after split

hichamboushaba committed:
Add dao functions for shipping labels

hichamboushaba committed:
Handle caching labels after fetching config

hichamboushaba committed:
Handle caching labels after fetching label status

hichamboushaba committed:
Cache labels after purchase request

hichamboushaba committed:
Allow restoring origin and address of labels when fetching status

hichamboushaba committed:
Get labels from the Room DB for the shipping labels form

hichamboushaba committed:
Get label from Room in the refund screen

hichamboushaba committed:
Get label from Room in the order details

hichamboushaba committed:
Remove the config data store

hichamboushaba committed:
Update unit tests

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>WooShippingDao</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/WooShippingLabelRepository.kt:
```diff
@@ -105,7 +116,24 @@ class WooShippingLabelRepository @Inject constructor(
-    ).asWooResult { response -> response.shippingLabel?.let { mapper(it) } }
+    ).also {
+        it.result?.shippingLabel?.let { labelDTO ->
+            // The API doesn't return the origin and destination addresses,
+            // we get it from the already cached labels in the database after the purchase request
+            val currentLabel = wooShippingDao.getLabel(
+                selectedSite.get().localId(),
+                LocalOrRemoteId.RemoteId(orderId),
+                labelId
+            )
+            val updatedLabel = mapper(labelDTO, site, orderId).copy(
+                originAddress = currentLabel?.originAddress,
+                destinationAddress = currentLabel?.destinationAddress
+            )
+            wooShippingDao.insertLabel(updatedLabel)
```
This is an improvement over what we had before. Before when getting the label from the API when we observe the status, we set the `originAddress` and `destinationAddress` to `null`, while now we cache the correct values in `purchaseShippingLabel`, and make sure to not override it here.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/GetShipments.kt:
```diff
@@ -64,55 +58,23 @@ class GetShipments @Inject constructor(
         }
     }
-
-    private fun mergeAddresses(
-        shipmentUIModelList: List<ShipmentUIModel>,
-        originAddresses: Map<String, OriginAddressDTO>?,
-        destinationAddresses: Map<String, DestinationAddressDTO>?
-    ): List<ShipmentUIModel> = shipmentUIModelList.map { shipmentUIModel ->
-        val remoteId = shipmentUIModel.remoteId ?: return@map shipmentUIModel
-        val key = getStoredDataKey(remoteId)
-
-        val updatedLabel = shipmentUIModel.label?.copy(
-            originAddress = originAddresses?.get(key)?.let { mapper(it) }
-                ?: shipmentUIModel.label.originAddress,
-            destinationAddress = destinationAddresses?.get(key)?.let { mapper(it) }
-                ?: shipmentUIModel.label.destinationAddress
-        )
-        shipmentUIModel.copy(label = updatedLabel)
-    }
```
This part is handled in the [mapper](https://github.com/woocommerce/woocommerce-android/blob/ed668fc9a66309daadeeb4d626d8e9ce73f68104/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/WooShippingNetworkingMapper.kt#L143-L186) now before caching the labels, so we don't need it here.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/purchased/ObserveShippingLabelStatus.kt:
```diff
@@ -42,56 +37,6 @@ class ObserveShippingLabelStatus @Inject constructor(
-            return if (alreadyHasSameLabel) {
-                currentOrderLabels.map {
-                    if (it.labelId == newLabel.labelId) {
-                        it.copy(
-                            created = newLabel.created?.time,
-                            status = newLabel.status,
-                            refundableAmount = newLabel.refundableAmount
-                        )
-                    } else {
-                        it
-                    }
-                }
-            } else {
-                // We map the new label to ShippingLabelDTO to save it into the Config Data Store. Not all properties
-                // are mapped, as only a subset is needed for the UI. The config will be updated again later when the
-                // config endpoint is called.
-                currentOrderLabels + ShippingLabelDTO(
-                    labelId = newLabel.labelId,
-                    tracking = newLabel.tracking,
-                    refundableAmount = newLabel.refundableAmount,
-                    status = newLabel.status,
-                    created = newLabel.created?.time,
-                    carrierId = newLabel.carrierId,
-                    currency = newLabel.currency,
-                    expiryDate = newLabel.expiryDate,
-                )
-            }
-        }
-
-        val currentConfig = shippingConfigDataStore.observeConfig(orderId).first() ?: return
-        val updatedConfig = currentConfig.copy(
-            shippingLabelData = currentConfig.shippingLabelData.copy(
-                currentOrderLabels = addPurchasedLabelToConfigDTO(
-                    currentConfig.shippingLabelData.currentOrderLabels ?: emptyList(),
-                    response
-                )
-            )
-        )
-        shippingConfigDataStore.saveConfig(orderId, updatedConfig)
-    }
```
This [logic](https://github.com/woocommerce/woocommerce-android/blob/f1fa2ad832807710d24cc2ff5be4bf56dff85366/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/networking/WooShippingLabelRepository.kt#L121-L132) in the repository replaces this part, and is easier to follow.

irfano commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/entity/WooShippingLabelEntity.kt:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    @TypeConverter
+    fun toAddressEntity(value: String?): WooShippingLabelEntity.Address? {
```
I encountered a crash on this function.
Crash logs:
```
Process: com.woocommerce.android, PID: 7623
                 java.lang.UnsupportedOperationException: JsonNull
                 	at com.google.gson.JsonElement.getAsString(JsonElement.java:252)
                 	at org.wordpress.android.fluxc.persistence.entity.WooShippingLabelConverters.toAddressEntity(WooShippingLabelEntity.kt:116)
                 	at org.wordpress.android.fluxc.persistence.dao.WooShippingDao_Impl$12.call(WooShippingDao_Impl.java:723)
                 	at org.wordpress.android.fluxc.persistence.dao.WooShippingDao_Impl$12.call(WooShippingDao_Impl.java:548)
                 	at androidx.room.CoroutinesRoom$Companion$execute$4$job$1.invokeSuspend(CoroutinesRoom.android.kt:69)
                 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
                 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1156)
                 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:651)
                 	at java.lang.Thread.run(Thread.java:1119)
                 	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@437a783, Dispatchers.Main.immediate]
```

To reproduce it, you can test order 1931 on our testing site.

hichamboushaba replied:
Great catch, this should be fixed by this commit a90c8ec

irfano reviewed (changes requested):
Great job, @hichamboushaba! Thanks for taking care of this and improving the shipping labels infrastructure.
It works as expected. I only noticed a crash and a difference in shipment sorting.

On the order detail screen, the sorting of packages is reversed in this PR. Previously, the sorting matched the web.

|Before|After|
|-|-|
|<img width="1344" height="2992" alt="before" src="https://github.com/user-attachments/assets/1b9a6d0f-4e22-4753-8227-51436c446b6d" />|<img width="1344" height="2992" alt="after" src="https://github.com/user-attachments/assets/a093824f-341c-4d0c-a137-b7e32d98f908" />|

hichamboushaba committed:
Fix JSON parsing issues

hichamboushaba committed:
Make sure to show only purchased labels in the order details

hichamboushaba commented:
Thanks @irfano for the review, I addressed the crash, and updated the sorting logic to match the previous behavior.
Even though, as part of the order details redesign, we will show the labels in the order of the shipments, so the order of the DB or the API won't matter anymore.

hichamboushaba committed:
Add a release note

hichamboushaba committed:
Merge branch 'trunk' into issue/WOOMOB-726-migrate-config-room

irfano committed:
Update RELEASE-NOTES.txt

irfano reviewed (approved):
Thank you! The issues have been fixed, and LGTM! I also corrected the PR number in the release notes before merging.
